### PR TITLE
Feature/detect outside variables strands

### LINKED ISSUE #8172: [p5.strands] Detect references to outside variables and give helpful feedback
### Increasing access

There are a number of subtle ways you can use variables in p5.strands that give no errors but subtly don't work. We can improve the experience for users to give them next steps for debugging.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [x] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [x] Friendly errors
- [ ] Other (specify if possible)

### Feature enhancement details

@ksen0 created this example and noted that if you were to use `if (mouseX/windowWidth < 0.3) {` instead of `if (value < 0.3) {` then it stops working, but without any errors.

```js
let testShader;
let pixelate;
let state_;


async function setup() {
  state_ = 0;
  createCanvas(windowWidth, windowHeight, WEBGL);
  testShader = baseMaterialShader().modify(() => {
    const value = uniformFloat(() => mouseX/windowWidth);
    const state = uniformFloat(() => state_);

    let time = uniformFloat(() => frameCount/5);
    getPixelInputs((inputs, canvasContent) => {
      let color = inputs.color;
      if (value < 0.3) {
        color = vec3(1.0, state*0.1, 0);
      } else if (value < 0.8) {
        color = vec3(state*0.1, 1, state*0.1);
      } else {
        color = vec3(state*0.1, 0, 1);
      }
      inputs.color = [color, 1.0];
      return inputs;
    });
  });
}

function draw() {
  background(0);
  shader(testShader);
  testShader.setUniform('value', mouseX/width)
  sphere(100);
}

function mousePressed(){
  state_ += 1;
  console.log(state_)
}
```

When something isn't a uniform, and isn't a value we specifically handle (see https://github.com/processing/p5.js/issues/8171; in the future we'll make `mouseX` and `windowWidth` available), it falls into one of two cases:
- It's a dynamic value, and should be turned into a uniform, e.g.: `const myValue = uniformFloat()` + `myShader.setUniform('myValue', something)` after applying the shader

- It's a static value, but it's a local variable that the strands callback can't access, and needs to be passed into an object as the second parameter of the `modify` function, e.g.: 

  ```js
  let myVal = 10
  const myShader = baseMaterialShader().modify(() => {
    getWorldInputs((inputs) => {
      inputs.position.x += myVal
      return inputs
    })
  }, { myVal }) // This is how to make a local variable accessible in strands
  ```

Maybe that second case isn't common enough to tell people about, so possibly just the first case is fine. Ideally we can detect usages of variables that aren't declared within the p5.strands callback, see if they're a known p5 global, and if not, output an example of making a uniform out of it?

### PR DESCRIPTION
**Pull request by Aayushdev18**
Resolves #8172

## Changes

This PR adds detection for outside variable references in `p5.strands` uniform initializers, providing helpful error messages to users when they reference variables that aren't declared in the strand context.

**What was implemented:**
- Created `detectOutsideVariableReferences()` function in `src/strands/strands_transpiler.js` that:
  - Uses Acorn AST parsing to analyze strand code
  - Collects all variable declarations within the strand context
  - Finds uniform initializer functions (e.g., `uniformFloat('color', () => ...)`)
  - Extracts variables referenced inside uniform initializers
  - Checks if referenced variables are declared in the strand scope
  - Returns helpful error messages for any undeclared variables
- Integrated detection to run before transpilation in `src/strands/p5.strands.js`
- Added unit tests in `test/unit/strands/strands_transpiler.js` with test cases for:
  - Detecting undeclared variables (e.g., `mouseX`, `windowWidth`)
  - No errors when variables are properly declared
  - Multiple undeclared variables
  - Edge cases

**Approach:**
Following @davepagurek's feedback, this implementation:
- Runs as a separate analysis phase before transpilation
- Reuses the transpiler's Acorn-based structure for AST parsing
- Analyzes variable declarations to build a set of in-scope variables
- Detects when uniform initializer functions reference variables not in that set
- Provides specific, actionable error messages to users

This is more robust than checking for hardcoded variable names, as it analyzes the actual code structure and works with any user-defined variables.

## Screenshots

N/A (infrastructure change - no visual impact)

## PR Checklist

- [x] Code follows p5.js coding conventions and practices
- [x] Code is tested and all tests pass
- [x] Documentation is updated (if applicable)
- [x] Pull request targets the correct branch (dev-2.0)
- [x] Changes are minimal and focused on the issue
- [x] No breaking changes

### CODE REVIEW

**Comment by Aayushdev18** [comment]
Hey @davepagurek! 

I've re-implemented the outside variable detection following your feedback:
- Uses AST analysis instead of hardcoded variable names
- Runs as a separate analysis phase before transpilation
- Properly handles the uniform function signature
- Includes unit tests
Could you take a look when you have time? Let me know if you'd like any adjustments!
Thanks!

File: test/unit/strands/strands_transpiler.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { detectOutsideVariableReferences } from '../../../src/strands/strands_transpiler.js';
+import { suite, test } from '../../../test/js/spec.js';
+
+suite('Strands Transpiler - Outside Variable Detection', function() {
+  test('should detect undeclared variable in uniform', function() {
+    // Simulate code that references mouseX (not declared in strand context)
+    const code = `
+      const myUniform = uniform('color', () => {
+        return mouseX; // mouseX is not declared
```
Comment (davepagurek) [inline test/unit/strands/strands_transpiler.js]: I think you might be checking for a different scenario than we need. In the callback of a uniform creation function, that's the *right* spot to have p5 globals. So this is ok:
```js
baseMaterialShader.modify(() => {
  const myUniform = uniformFloat(() => mouseX)
  getWorldPosition((inputs) => {
    inputs.position.x += myUniform
    return inputs
  })
})
```
...but this should show a warning:
```js
baseMaterialShader.modify(() => {
  getWorldPosition((inputs) => {
    inputs.position.x += mouseX
    return inputs
  })
})
```

Additionally, just `uniform()` on its own isn't a p5 function, it will always have a type after it, like `uniformFloat` or `uniformVec2`.

File: src/strands/strands_transpiler.js [hunk 2]
```diff
@@ -863,7 +863,120 @@ const ASTCallbacks = {
...
+            if (scopeChain.length > 0) {
+              scopeChain[scopeChain.length - 1].add(decl.id.name);
+            }
+          }
+        }
+      }
+      
+      // Close block scope when exiting
+      if (node.type === 'BlockStatement' && scopeChain.length > 1) {
+        scopeChain.pop();
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js]: Right now it looks like `scopeChain` doesn't get used really -- it gets added to and then popped off the array before anything reads from it. Maybe consider doing the check all in one step, so that when you encounter a variable identifier, you can check to see if it should be visible in the current scope?

**Comment by Aayushdev18** [comment]
@davepagurek Fixed! Now checking strand code (not uniform callbacks) and using the simpler two-pass approach. Tests updated. Ready for another look when you have time.

File: test/unit/strands/strands_transpiler.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    `;
+    
+    const errors = detectOutsideVariableReferences(code);
+    assert.ok(errors.length > 0, 'Should detect error');
+    assert.ok(errors.some(e => e.variable === 'mouseX'), 'Should detect mouseX');
+  });
+
+  test('should not error when variable is declared', function() {
+    const code = `
+      baseMaterialShader.modify(() => {
```
Comment (davepagurek) [inline test/unit/strands/strands_transpiler.js]: I think this is testing something different from what the function would actually receive. This looks like code that would be found in a p5 sketch, but just the contents of the `modify` callback is sent into the your function normally.

I think a good next step would be to try to do a full integration test by running `yarn dev:global` and editing the sketch in `preview/global`. If that works, you can also show the maintainers by building the library with `npm run build`, and then uploading `lib/p5.js` to a p5 web editor sketch so that others can run the same test. I suspect doing that will reveal some of the problems with the implementation more quickly than in review (but feel free to reach out if you run into issues while doing so!)

File: src/strands/strands_transpiler.js [hunk 4]
```diff
@@ -863,7 +863,89 @@ const ASTCallbacks = {
...
+        if (ignoreNames.includes(varName)) return;
+        
+        // Skip if it's a property access (obj.prop)
+        const isProperty = ancestors.some(anc => 
+          anc.type === 'MemberExpression' && anc.property === node
+        );
+        if (isProperty) return;
+        
+        // Skip if it's a function parameter
+        const isParam = ancestors.some(anc =>
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js]: What is this case aiming to catch? `inputs` in e.g. `getWorldInputs((inputs) => { ... })`?

File: src/strands/strands_transpiler.js [hunk 5]
```diff
@@ -863,7 +863,89 @@ const ASTCallbacks = {
...
+        // Check if we're inside a uniform callback (OK to access outer scope)
+        const inUniformCallback = ancestors.some(anc => 
+          anc.type === 'CallExpression' &&
+          anc.callee.type === 'Identifier' &&
+          anc.callee.name.startsWith('uniform')
+        );
+        if (inUniformCallback) return; // Allow outer scope access in uniform callbacks
+        
+        // Check if this variable is declared anywhere in the strand
+        if (!declaredVars.has(varName)) {
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js:1000]: I think what you had before with pushing/popping scopes was not a bad approach -- currently, it looks like this will pass, when it should not:
```js
baseMaterialShader().modify(() => {
  getWorldInputs(() => {
    let i = 1
    inputs.position.x += i // OK, i is in scope
    return inputs
  })
  getFinalColor((color) => {
    color.r += i // Not OK, i is not in scope
    return color
  })
})
```

**Comment by Aayushdev18** [comment]
@davepagurek Thanks for catching those issues! Updated the tests to pass just the function body, and the scope checking should now properly exclude sibling scopes using the ancestor chain.
Going to run the integration tests as you suggested - expect that will surface any issues I missed. Thanks for the detailed guidance! 💯

**Comment by Aayushdev18** [comment]
Hi @davepagurek , just following up, it’s been over a week since this PR was opened. Could you please approve the pending workflows or review it when you get a chance? Thanks!

File: src/strands/strands_transpiler.js [hunk 6]
```diff
@@ -920,7 +920,100 @@ const ASTCallbacks = {
...
+    // Second pass: check identifier references
+    ancestor(ast, {
+      Identifier(node, state, ancestors) {
+        const varName = node.name;
+        
+        // Skip built-ins and p5.strands functions
+        const ignoreNames = [
+          '__p5', 'p5', 'window', 'global', 'undefined', 'null', 'this', 'arguments',
+          // p5.strands built-in functions
+          'getWorldPosition', 'getWorldNormal', 'getWorldTangent', 'getWorldBinormal',
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js:957]: This approach is currently hard-coding these identifiers, but they can be almost anything (a custom shader can create its own hooks for other users), so we'll need a different approach. A number of the identifiers in this list don't appear anywhere in any default shaders either, such as `getWorldBinormal`, and feel like AI hallucinations. Given the complex nature of this task and the system it's interacting with, I think we really need to have a deeper understanding of what the approach here should be before proceeding with writing code, otherwise it will take a significant amount of back-and-forth in review. Is that something you're up for? If so, we can discuss an approach to how to tackle this (and feel free to reach out in the p5 discord, I'm happy to explain in a less asynchronous way there!) If not, also no worries -- this is a complex issue, and there are others that you can tackle, and we can open this one up to other contributors.
